### ISSUE TITLE
Feature Request: PolymorphicJsonAdapter flag to suppress automatic inclusion of label in toJson()

### ISSUE THREAD

**Issue by danielesegato** [square/moshi#874]
I've a series of classes that contains a `type` attribute (string). The Polymorphic adapter is great to parse and serialize those, but I sometimes need to use a specific adapter to do the same thing.

This means I need the autogenerated adapter for the actual type to write the `type` attribute (I cannot mark it as `@Transient`) but I do not want the Polymorphic adapter to produce a double `type` attribute.

```
interface DataItem {
  val type: String
}

@JsonClass(generateAdapter = true)
data class Text(
  val text: String
): DataItem {
      override val type: String = "text"
}

@JsonClass(generateAdapter = true)
data class Image(
  val url: String,
  val w: Int,
  val h: Int
): DataItem {
    override val type: String = "image"
}
```

If I setup the Polymorphic adapter for these types:

```
PolymorphicJsonAdapterFactory.of(DataItem::class.java, "type")
            .withSubtype(Text::class.java, "text")
            .withSubtype(Image::class.java, "image")
```

when I serialize I get:

```
{
    "type": "text",
    "type": "text",
    "text": "foo"
}
```

if I mark the `type` field as `@Transient` and then use the generated `Text` adapter directly I get

```
{
   "text": "foo"
}
```

I'd like to obtain
```
{
    "type": "text"
    "text": "foo"
}
```

in both cases.

Something like this?
```
PolymorphicJsonAdapterFactory.of(DataItem::class.java, "type")
            .withSubtype(Text::class.java, "text")
            .withSubtype(Image::class.java, "image")
            .skipLabelKeyInSerialization()
```

And the code in the PolymorphicAdapter could be changed to
```
    @Override public void toJson(JsonWriter writer, Object value) throws IOException {
      Class<?> type = value.getClass();
      int labelIndex = subtypes.indexOf(type);
      // ...
      JsonAdapter<Object> adapter = jsonAdapters.get(labelIndex);
      writer.beginObject();
      if (!skipLabelKey) {
          writer.name(labelKey).value(labels.get(labelIndex));
      }
      int flattenToken = writer.beginFlatten();
      adapter.toJson(writer, value);
      writer.endFlatten(flattenToken);
      writer.endObject();
    }
```

**Comment by abhishekBansal** [comment]
+1 
In my case I already have labels in my class, `toJson()` should check if type key is already there. If its there then it should avoid duplicating it.

**Comment by abhishekBansal** [comment]
How will @Transient behave at the time serialisation?

**Comment by ZacSweers** [comment]
It won't be serialized. However, you might hit the issue in #858 then in this case

**Comment by ZacSweers** [comment]
I've put up a proposal implementation to resolve this in #925

**Comment by danielesegato** [comment]
@ZacSweers I can't mark it as `@Transient`

in my example I had 2 objects: `Text` and `Image` and the base type was `DataItem`

I've needs to serialize and deserialize to/from Json:

- directly using the generated `TextJsonAdapter` or `ImageJsonAdapter`
- through the base polymorphic adapter of `DataItem`

if I mark the `type` field `@Transient` the two generated `TextJsonAdapter` and `ImageJsonAdapter` will not contain a `type` attribute when serialized and the polymorphic adapter will work as I expect.

If I leave the `type` field without `@Transient` the two generated `TextJsonAdapter` and `ImageJsonAdapter` will correctly serialize it but when serializing with the polymorphic adapter I'll get a double `type` attribute.

In my code I implemented my own version of the polymorphic adapter that let me specify I don't want it to serialize the `type` in a way similar as the one proposed above.

I may be mistaken but I don't think you addressed this issue with your merge request, you are addressing another problem with the polymorphic adapter (it's inability to work correctly when `failOnUnknown()` is set).

**Comment by ZacSweers** [comment]
The PR proposal addresses both that issue and this one, please look again. Specifically the following commits:

https://github.com/square/moshi/pull/925/commits/9d2e0dd1dd9b6e9db6e11004dbb7be3a34d67d9a
https://github.com/square/moshi/pull/925/commits/29994a1fbe0ed49e3dfd24b3495cd87643d00b71
https://github.com/square/moshi/pull/925/commits/fd8b756d2689a2da34d28b0c19a85012a1d65e50

**Comment by danielesegato** [comment]
Yes, you are right, this does make the polymorphic adapter works as I want.
But the skipping api added to the writer looks a bit too linked to the Polymorphic adapter needs to me.

I think it would be better if the solution could be more localized to the adapter or be put in the hands of the users (for example by creating a new annotation to hint adapters on what to do with a variable).

This is my personal opinion tho' :)

**Comment by ZacSweers** [comment]
I'm not personally a fan of another annotation for this, nor any configuration for it really. The ideal case is that it should Just Work. The specific use-case in that PR is definitely for PolymorphicJsonAdapterFactory, but we've made general-use APIs before based on needs that custom adapters exposed.

**Comment by eneim** [comment]
Hi, recently I have the same issue with this and want to ask if there is any future plan fixing/closing this issue? Thanks.

**Comment by mortenholmgaard** [comment]
@swankjesse @ZacSweers This issue don't seem to be fixed or have a working workaround. The poposed transient fix does not seem to work, at least not for an abstract baseclass.
Why was the PR abandoned, without another fix(as I could find)? https://github.com/square/moshi/pull/925

**Comment by alizeec** [comment]
Hello! 
I have the opposite issue, on my side I don't want the "type" in the json at all, but even with the `@Transient` annotation I have it

```
interface ApiBodyRequestValue {
    val type: String

    companion object {
        val adapterFactory: PolymorphicJsonAdapterFactory<ApiBodyRequestValue> =
            PolymorphicJsonAdapterFactory.of(ApiBodyRequestValue::class.java, "type")
                .withSubtype(
                    ValueRate::class.java,
                    "rate",
                )
                .withSubtype(
                    ValueComment::class.java,
                    "comment",
                )
    }
}

@JsonClass(generateAdapter = true)
data class ValueRate(val rate: Float) :
    ApiBodyRequestValue {
    @Transient
    override val type = "rate"
}

@JsonClass(generateAdapter = true)
data class ValueComment(val comment: String) :
    ApiBodyRequestValue {
    @Transient
    override val type =  "comment"
}
```

gives me 

```
value: {
 rate: 4
 type: "rate"
}
``` 

Do you have any idea why ?

Thanks
